### ISSUE TITLE
Customizable project structure (platforms)

### ISSUE THREAD

**Issue by passsy** [flutter/flutter#71400]
## Use case

Flutter supports multiple platforms, currently 6 (ios, android, windows, linux, macos, web).
 
A project with all platforms enabled feels unorganized to me. The platform folders are placed "randomly" between project-related folders. In many cases they contain no to minimal code. Thus I don't think they are important enough to be placed top-level. 

I'd prefer when all platform-specific sub-projects would be combined into a single `platforms` folder.

<table>
<tr>
<td>
Before top-level view
</td>
<td>
Proposed
</td>
<td>
Proposed top-level view
</td>
<tr>
<td>

```
flutter_project
├── .github
├── .vscode
├── android
├── assets
├── docs
├── example
├── ios
├── lib
├── linux
├── macos
├── test
├── driver_test
├── web
├── windows
├── pubspec.yaml
├── pubspec.lock
│
```
</td>
<td>

```
flutter_project
├── .github
├── .vscode
├── assets
├── docs
├── example
├── lib
├── platforms
│   ├── android
│   ├── linux
│   ├── macos
│   ├── web
│   ├── windows
│   ├── macos
├── test
├── test_driver
├── pubspec.yaml
├── pubspec.lock
```
</td>
<td>

```
flutter_project
├── .github
├── .vscode
├── assets
├── docs
├── example
├── lib
├── platforms
├── test
├── test_driver
├── pubspec.yaml
├── pubspec.lock






```
</td>
</table>

## Custom platform paths

Instead of forcing all existing and future project to migrate to the new structure, it has to be opt-in either hardcoded or customizable

### Hardcoded

When the Flutter team decides that a `platforms` directory is a brilliant idea and adopted as an alternative project layout it could be hardcoded in the `flutter_tool` supporting the old and new layout.

### Customizable

One doesn't fit all. Allowing customizations would allow everyone to place platform folders where it makes sense in their repository. It could be especially for mono repos with very different structures.

Similar to assets, the `pubspec.yaml` could contain a section to platform folders.

```
flutter:
  platforms:
    android:
      - platforms/android
    ios:
      - ../../some-ios-project
```

**Comment by lukepighetti** [comment]
I don't think it should be customizable, but I think the proposed new folder layout makes sense.

**Comment by evandrmb** [comment]
The proposed model seems very clean, I liked it. Can't think, right now, why someone would want to customize it.

**Comment by YazeedAlKhalaf** [comment]
It is really so clean 🙌🚀, totally agree!

**Comment by 7Mendez** [comment]
It's great idea 💡.

**Comment by searchy2** [comment]
Would this introduce any breaking changes to Add 2 App?

**Comment by JMario1** [comment]
Love it

**Comment by SergioEric** [comment]
What about having optional platforms?
When we only want to develop something for an exclusive platform!

**Comment by RegNex** [comment]
I will definitely prefer the proposed. It looks clean and well structured.

**Comment by jonahwilliams** [comment]
I think the proposed new layout looks fine, and while I wouldn't personally use a configurable project directory I could definitely understand cases where that would be useful.

The problem is that making this change is non-trivial for a number of reasons:
  * We would need to indefinitely support both structures, because the changes seems large enough that I don't think any sort of auto-migration would be feasible. I could not ever justify a purely cosmetic breaking change.
  * We would need to support correctly detecting in the tool at runtime which configuration is used. This is going to be tricky to get 100% correct, and there will be tons of bugs filled where users have gotten their projects into a weird state and can't build.
  * All of the flutter.dev documentation, tutorial, code labs, blogs posts (google and non-google) will suddenly be out of date. For example: https://flutter.dev/docs/deployment/android. In the worst case, a new user will only stumble upon outdated blog posts and get frustrated and give up on flutter entirely.
  * All of the IDEs and tools (google and non-google) that may depend on the project structure in a particular way would also need to be updated with the same logic used in flutter_tool. Discovery will be a difficult problem: we don't know what internal tools or personal scripts users have written to work with flutter apps.
  * Plus, as in all software projects, a bunch of surprises along the way.

**Comment by escamoteur** [comment]
Why wouldn't it be possible to auto migrate this?

**Comment by passsy** [comment]
Some platform build scripts reference `../` as "flutter_root". While this could be migrated (still hard) `flutter_tool` can't detect all locations where a custom script references `/${platform}. Those scripts might be placed outside of the project, like CI/CD scripts where `cd android && ./gradlew someTask` is hardcoded and would break.

Detection for 3rd party software will be tricky. Either they have to check two locations (`/windows` or `/platforms/windows`) or parse the `pubspec.yaml` to get the correct directory. To prevent both, `flutter_tool` should offer a new command printing the path for a platform.

```
$ flutter platform linux --path
/Users/pascal/project/my_flutter_project/platforms/linux
```

**Comment by lukepighetti** [comment]
if project_folder contains "iOS" | "android" | "web" | "linux" | "macOS" | "windows"
move (& simlink?) folders "iOS" | "android" | "web" | "linux" | "macOS" | "windows" to "./platforms"

**Comment by jonahwilliams** [comment]
Copying the folders around would not be hard, no. The problem is that users may have their own infrastructure that refers to parts of the Gradle/Xcode projects directory or additions to these project that make assumptions about the directory they are in relative to the flutter application.  The flutter tool will not be able to detect this, which leaves it up to each developer to fix on their own.

**Comment by escamoteur** [comment]
yeah, maybe we could setup symbolic links for transfer time?
On the other side if we get even more platforms it will get worse. unfortunately because the don't start with `platform_` or just a `_` the platform folders are scattered throughout the project folder.

**Comment by lukepighetti** [comment]
I can't think of a reason why we can't have breaking folder layout changes on a major flutter release.

**Comment by lukepighetti** [comment]
@passsy would you consider dropping the customization from this issue? It seems like 90% of the issue is the `projects` folder.

**Comment by YazeedAlKhalaf** [comment]
I believe people will have to change the folder location and it will be easy. After changing the creation tool way of creating things and making everything work when we create platforms in one location. it will be a matter of"
1) create a folder with the name platforms
2) move all platform folders to the platform folder

And then the folders will look good, and adding more platforms will be way easier 🚀

**Comment by tananga** [comment]
Very good idea !

**Comment by stuartmorgan** [comment]
See https://github.com/flutter/flutter/issues/34536#issuecomment-1400512538 for a partial list of the practical downsides of the proposed change. It covers a number of the same things as [Jonah's comment above](https://github.com/flutter/flutter/issues/71400#issuecomment-736180298), but has some others as well.

**Comment by polina-c** [comment]
If we ever do this migration, it would make sense to make test structure cleaner:

```
flutter_project
├── .github
├── .vscode
├── assets
├── docs
├── example
├── lib
├── platforms
│   ├── android
│   ├── linux
│   ├── macos
│   ├── web
│   ├── windows
│   ├── macos
├── test
│   ├── driver
│   ├── integration
│   ├── unit
├── pubspec.yaml
├── pubspec.lock
```

**Comment by AnubisAwooo** [comment]
![image](https://github.com/flutter/flutter/assets/46840787/1b9f4de0-783b-4109-a480-8d477f8c40fa)

A temporary solution. 😂😂

**Comment by AnubisAwooo** [comment]
A temporary solution. 😂😂

反正要被笑死😄

![266766644-8416999d-309a-4797-95a2-5b7c41921302](https://github.com/flutter/flutter/assets/46840787/4c168e29-24fc-49a6-96be-8fd4dee53363)

**Comment by cedvdb** [comment]
> Will all projects be force-migrated, or will flutter need to support both structures forever?

You could have a grace period with a `platform-root: ` option in `pubspec.yml`. It would currently have the value `./` but at some point be migrated to `./platforms`, once enough packages use it.

> What would the plan be for updating all of our first-party documentation, which refers to the old paths everywhere ?

That's a resource issue, not much we can suggest beside the obvious: the documentation will have to be updated from "the android folder" to the "$platforms-root/android folder", or whatever method is used.

> What happens to all of the third-party tooling out there that expects the current structure (e.g., the packages that streamline updating app icons or names across all platforms)?

They can start migrating to `$platform-root` without impact.

Is this naive ?

**Comment by cedvdb** [comment]
@stuartmorgan Your last comment doesn't really apply here (it did apply where it was posted).
 
 > How would you propose handling all of that?

The proposal is to have the platform directory inside pubspec and if the information is not present, default to the root directory (like it is today).

```yaml
  platform-directory: ./platforms
``` 

The changes that I can foresee are:

  - Existing flutter projects are not impacted - by the definition above, they already use the default -
  - Flutter repositories will have to be refactored to use that information (or fallback to the default). Searching inside the flutter project for "android/" I see 792 results and 70% of them are not relevant. That still be quite a task but that looks pretty  manageable. 
 

 > Will all projects be force-migrated, or will flutter need to support both structures forever?

None of the above. Existing projects do not have a `platform-directory` parameter in their pubspec.yaml, so they use the default, which by this proposal, is the same as today.

 > Will we force-modify everyone's .gitignore files? Because sufficiently old projects have all their platform-specific ignores in a singe top-level .gitignore, so if we don't the git state of old projects will be suddenly wrong (but if we do, we have to auto-modify a file that has potentially arbitrary user changes).

Again this does not really apply to the proposal, I'll skip the rest of the "rebutals / questions"  that were primarily aimed at the other proposal.

 > What, if anything, will be done about the inconsistency with plugins, which also have all of these same directories, but can't be force-migrated?

This seem a bit out of scope, from an end user standpoint, plugins can stay the same as they are today, unless there is a technical reason it can't.

**Comment by stuartmorgan** [comment]
> The proposal is to have the platform directory inside pubspec and when the information is not present, default to the root directory (like it is today).

There are two different proposals in the original issue: "Hardcoded" and "Customizable". Your comments seem to assume the "Customizable" option is what would be implemented. Many duplicates of this issue have proposed the hard-coded version instead.

**Comment by cedvdb** [comment]
> There are two different proposals in the original issue: "Hardcoded" and "Customizable". Your comments seem to assume the "Customizable" option is what would be implemented.

Yes, the "customizable" option seems to suffer from less drawbacks at first glance at least. Even if it's customizable it can be customizable upon two paths only (only the root folder or the platforms folder). The main advantage is that this is not a breaking changes for existing projects. 

> This would add a large amount of permanent complexity to the documentation, because every instance of a path that involves a platform would need this caveat.

Yes indeed, I can't think of something better though. 


Just to be clear, your stance is that this is not worth it ?
